app/controllers/materials_controller.rb

```diff
@@ -0,0 +1,221 @@
+require 'rubygems'
+require 'zip'
+
+class MaterialsController < ApplicationController
+
+  before_action :set_and_authorize_event
+
+  # POST /events/1/upload_material
+  def upload_material
+    Dir.mkdir(@event.material_path) unless Dir.exists?(@event.material_path)
+
+    path = params[:path].to_s
+    file = params[:file_upload]
+
+    if !is_file?(file)
+      redirect_to event_path(@event), alert: t("events.material_area.no_file_given") and return
+    elsif invalid_pathname?(path) || invalid_filename?(file.original_filename)
+      redirect_to event_path(@event), alert: t("events.material_area.invalid_path_given") and return
+    elsif !File.directory?(File.join(@event.material_path, path))
+      redirect_to event_path(@event), alert: t("events.material_area.download_file_not_found") and return
+    end
+
+    file_path = File.join(@event.material_path, path, file.original_filename)
```

Comment:
Since the `original_filename` is user controlled, I am not 100% sure it can't contain `../`. Something to verify..